File: 19116040-08/19116040-05/Source Code/P2.cpp
#include <iostream>
using namespace std;

class linkedList{
    private:
        struct node{
            int vertex1 = -1;
            int vertex2 = -1;
            int weight = -1;
            node *next;
            bool present = false;
        }*head;

        int numberOfEdges;
        int numberOfNodes;

    public:
        linkedList();
        void addNumberOfEdges(int n);
        void calculateNumberOfNodes();
        void insert(int v1, int v2, int w);
        int findRoot(int a, int *container);
        bool find(node *q, int *container);
        void doUnion(node *q, int *container);
        void kruskalAlgorithm();
        int totalEdgeWidth();
        void printSpanningTree();
        void printDotFile();
};

linkedList::linkedList(){
    head = NULL;
}

void linkedList::addNumberOfEdges(int n){
    numberOfEdges = n;
}

void linkedList::calculateNumberOfNodes(){
    int* arr;
    arr = (int*)calloc(numberOfEdges + 1, sizeof(int));
    int count = 0;
    node *q;
    q = head;
    while(q != NULL){
        int v1 = q->vertex1;
        int v2 = q->vertex2;
        if(arr[v1 - 1] == 0){
            arr[v1 - 1] = 1;
            count = count + 1;
        }
        if(arr[v2 - 1] == 0){
            arr[v2 - 1] = 1;
            count = count + 1;
        }
        q = q->next;
    }

    numberOfNodes = count;
    free(arr);
}

void linkedList::insert(int v1, int v2, int w){
    node *s;
    if(head == NULL){
        head = new node;
        head->vertex1 = v1;
        head->vertex2 = v2;
        head->weight = w;
        head->next = NULL;
    }else{
        s = new node;
        s->vertex1 = v1;
        s->vertex2 = v2;
        s->weight = w;
        node *p, *q;
        p = head;
        q = head;
        while(q != NULL){
            if(q->weight > w){
                break;
            }
            p = q;
            q = q->next;
        }
        if(p == q){
            s->next = head;
            head = s;
        }else{
            p->next = s;
            s->next = q;
        }
    }
}

int linkedList::findRoot(int a, int *container){
    int s = container[a - 1];
    if(s >= 0){
        a = findRoot(s, container);
    }
    return(a);
}

bool linkedList::find(node *q, int *container){
    int v1 = q->vertex1;
    int v2 = q->vertex2;

    int r1 = container[v1 - 1];
    int r2 = container[v2 - 1];

    if(r1 != 0 && r2 != 0){
        int x = findRoot(v1, container);
        int y = findRoot(v2, container);
        if(x == y){
            return(false);
        }
    }
    return(true);
}

void linkedList::doUnion(node *q, int *container){
    int v1 = q->vertex1;
    int v2 = q->vertex2;
    q->present = true;

    int r1 = container[v1 - 1];
    int r2 = container[v2 - 1];

    if(r1 == 0 && r2 == 0){
        container[v1 - 1] = -2;
        container[v2 - 1] = v1;
    }else if(r1 == 0){
        int x = findRoot(v2, container);
        container[v1 - 1] = x;
        container[x - 1] = container[x - 1] - 1;
    }else if(r2 == 0){
        int x = findRoot(v1, container);
        container[v2 - 1] = x;
        container[x - 1] = container[x - 1] - 1;
    }else{
        int x = findRoot(v1, container);
        int y = findRoot(v2, container);
        if(container[x-1] < container[y-1]){
            int z = container[y-1];
            container[y-1] = x;
            container[x-1] = container[x-1] + z;
        }else{
            int z = container[x-1];
            container[x-1] = y;
            container[y-1] = container[y-1] + z;
        }
    }
}

void linkedList::kruskalAlgorithm(){
    int* container;
    container = (int*)calloc(numberOfNodes + 1, sizeof(int));

    node *q;
    q = head;
    while(q != NULL){
        bool x = find(q, container);
        if(x == true){
            doUnion(q, container);
        }
        q = q->next;
    }

    free(container);
}

int linkedList::totalEdgeWidth(){
    node *q;
    q  = head;
    int w = 0;
    while(q != NULL){
        if(q->present == true){
            w = w + q->weight;
        }
        q = q->next;
    }
    return(w);
}

void linkedList::printSpanningTree(){
    cout << "// Node1  Node2  Weight\n";
    node *q;
    q = head;
    while(q != NULL){
        if(q->present == true){
            int v1 = q->vertex1;
            int v2 = q->vertex2;
            int w = q->weight;
            cout << "// " << v1 << "      " << v2 << "      " << w << "\n";
        }
        q = q->next;
    }
}

void linkedList::printDotFile(){
    cout << "graph{\n";
    node *q;
    q = head;
    while(q != NULL){
        if(q->present == true){
            int v1 = q->vertex1;
            int v2 = q->vertex2;
            int w = q->weight;
            cout << "\t" << v1 << " -- " << v2 << " [label = \"" << w << "\"];\n";
        }
        q = q->next;
    }
    cout << "}";
}

int main() {
    linkedList l1;

    int n = 0;
    int v1, v2, w;
    while(cin >> v1 >> v2 >> w){
        l1.insert(v1, v2, w);
        n = n + 1;
    }

    l1.addNumberOfEdges(n);
    l1.calculateNumberOfNodes();

    l1.kruskalAlgorithm();
    
    l1.printSpanningTree();
    w = l1.totalEdgeWidth();
    cout << "//Total edge-weight of the MST is: " << w << "\n";

    cout << "//Dot file output is: \n";
    l1.printDotFile();
}

File: 19116040-11/Problem 2/P2.java
import java.util.Scanner;

class linkedList{
    class node{
        int vertex1;
        int vertex2;
        int weight;
        node next;
        boolean isPresent = false;
    }

    node head;
    int setContainer[];

    public linkedList(int k){
        head = null;
        setContainer = new int[k];
    }

    public void insert(int v1, int v2, int w){
        node n = new node();
        n.vertex1 = v1;
        n.vertex2 = v2;
        n.weight = w;

        if(head == null){
            n.next = null;
            head = n;
        }else{
            node p, q;
            p = head;
            q = head;
            while(q != null){
                if(q.weight > n.weight){
                    break;
                }else{
                    p = q;
                    q = q.next;
                }
            }
            if(p == q){
                n.next = q;
                head = n;
            }else{
                p.next = n;
                n.next = q;
            }
        }
    }

    public boolean find(node n){
        int v1 = n.vertex1;
        int v2 = n.vertex2;

        int r1 = rootElement(v1);
        int r2 = rootElement(v2);

        if(r1 == r2){
            return(false);
        }else{
            return(true);
        }
    }

    public int rootElement(int v){
        if(setContainer[v - 1] <= 0){
            return(v);
        }else{
            int a = setContainer[v - 1];
            int b = rootElement(a);
            return(b);
        }
    }

    public void union(node n){
        int v1 = n.vertex1;
        int v2 = n.vertex2;

        int root1 = rootElement(v1);
        int root2 = rootElement(v2);

        int r1 = setContainer[root1 - 1];
        int r2 = setContainer[root2 - 1];

        if(r1 == 0 && r2 == 0){
            setContainer[v1 - 1] = -2;
            setContainer[v2 - 1] = v1;
        }else if(r1 == 0){
            setContainer[v1 - 1] = root2;
            setContainer[root2 - 1] = setContainer[root2 - 1] - 1;
        }else if(r2 == 0){
            setContainer[v2 - 1] = root1;
            setContainer[root1 - 1] = setContainer[root1 - 1] - 1;
        }else{
            if(r1 < r2){
                setContainer[root1 - 1] = setContainer[root1 - 1] + setContainer[root2 - 1];
                setContainer[root2 - 1] = root1;
            }else{
                setContainer[root2 - 1] = setContainer[root2 - 1] + setContainer[root1 - 1];
                setContainer[root1 - 1] = root2;
            }
        }
    }

    public void kruskalAlgorithm(){
        node n = head;
        while(n != null){
            if(find(n)){
                union(n);
                n.isPresent = true;
            }
            n = n.next;
        }

        printOutput();
    }

    public void printOutput(){
        node n = head;
        while(n != null){
            if(n.isPresent){
                int v1 = n.vertex1;
                int v2 = n.vertex2;
                int w = n.weight;
                if(v1 < v2){
                    System.out.println(v1 + " " + v2 + " " + w);
                }else{
                    System.out.println(v2 + " " + v1 + " " + w);
                }
            }
            n = n.next;
        }
    }
}

class P2{
    public static void main(String args[]) {
        Scanner sc = new Scanner(System.in);
        int t = sc.nextInt();
        int i;
        for(i = 0; i < t; i++){
            int vertices = sc.nextInt();
            int edges = sc.nextInt();
            int j;

            linkedList list = new linkedList(vertices);
            for(j = 0; j < edges; j++){
                int v1 = sc.nextInt();
                int v2 = sc.nextInt();
                int w = sc.nextInt();

                list.insert(v1, v2, w);
            }

            list.kruskalAlgorithm();
        }
    }
}